// angular-2-dropdown-multiselect/src/multiselect-dropdown.js
/*
 * Angular 2 Dropdown Multiselect for Bootstrap
 *
 * Simon Lindh
 * https://github.com/softsimon/angular-2-dropdown-multiselect
 */
System.register(["@angular/core", "@angular/common", "@angular/forms"], function (exports_1, context_1) {
    "use strict";
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __moduleName = context_1 && context_1.id;
    var core_1, common_1, forms_1, MULTISELECT_VALUE_ACCESSOR, MultiSelectSearchFilter, MultiselectDropdown, MultiselectDropdownModule;
    return {
        setters: [
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (common_1_1) {
                common_1 = common_1_1;
            },
            function (forms_1_1) {
                forms_1 = forms_1_1;
            }
        ],
        execute: function () {/*
             * Angular 2 Dropdown Multiselect for Bootstrap
             *
             * Simon Lindh
             * https://github.com/softsimon/angular-2-dropdown-multiselect
             */
            MULTISELECT_VALUE_ACCESSOR = {
                provide: forms_1.NG_VALUE_ACCESSOR,
                useExisting: core_1.forwardRef(function () { return MultiselectDropdown; }),
                multi: true
            };
            MultiSelectSearchFilter = (function () {
                function MultiSelectSearchFilter() {
                }
                MultiSelectSearchFilter.prototype.transform = function (options, args) {
                    return options.filter(function (option) {
                        return option.name
                            .toLowerCase()
                            .indexOf((args || '').toLowerCase()) > -1;
                    });
                };
                return MultiSelectSearchFilter;
            }());
            MultiSelectSearchFilter = __decorate([
                core_1.Pipe({
                    name: 'searchFilter'
                })
            ], MultiSelectSearchFilter);
            exports_1("MultiSelectSearchFilter", MultiSelectSearchFilter);
            MultiselectDropdown = (function () {
                function MultiselectDropdown(element, differs) {
                    this.element = element;
                    this.differs = differs;
                    this.selectionLimitReached = new core_1.EventEmitter();
                    this.dropdownClosed = new core_1.EventEmitter();
                    this.numSelected = 0;
                    this.isVisible = false;
                    this.searchFilterText = '';
                    this.defaultSettings = {
                        pullRight: false,
                        enableSearch: false,
                        checkedStyle: 'checkboxes',
                        buttonClasses: 'btn btn-default btn-secondary',
                        selectionLimit: 0,
                        closeOnSelect: false,
                        autoUnselect: false,
                        showCheckAll: false,
                        showUncheckAll: false,
                        dynamicTitleMaxItems: 3,
                        maxHeight: '300px',
                    };
                    this.defaultTexts = {
                        checkAll: 'Check all',
                        uncheckAll: 'Uncheck all',
                        checked: 'checked',
                        checkedPlural: 'checked',
                        searchPlaceholder: 'Search...',
                        defaultTitle: 'Select',
                    };
                    this.onModelChange = function (_) { };
                    this.onModelTouched = function () { };
                    this.differ = differs.find([]).create(null);
                }
                MultiselectDropdown.prototype.onClick = function (target) {
                    var parentFound = false;
                    while (target != null && !parentFound) {
                        if (target === this.element.nativeElement) {
                            parentFound = true;
                        }
                        target = target.parentElement;
                    }
                    if (!parentFound) {
                        this.isVisible = false;
                    }
                };
                MultiselectDropdown.prototype.ngOnInit = function () {
                    this.settings = Object.assign(this.defaultSettings, this.settings);
                    this.texts = Object.assign(this.defaultTexts, this.texts);
                    this.title = this.texts.defaultTitle;
                };
                MultiselectDropdown.prototype.writeValue = function (value) {
                    if (value !== undefined) {
                        this.model = value;
                    }
                };
                MultiselectDropdown.prototype.registerOnChange = function (fn) {
                    this.onModelChange = fn;
                };
                MultiselectDropdown.prototype.registerOnTouched = function (fn) {
                    this.onModelTouched = fn;
                };
                MultiselectDropdown.prototype.ngDoCheck = function () {
                    var changes = this.differ.diff(this.model);
                    if (changes) {
                        this.updateNumSelected();
                        this.updateTitle();
                    }
                };
                MultiselectDropdown.prototype.clearSearch = function () {
                    this.searchFilterText = '';
                };
                MultiselectDropdown.prototype.toggleDropdown = function () {
                    this.isVisible = !this.isVisible;
                    if (!this.isVisible) {
                        this.dropdownClosed.emit();
                    }
                };
                MultiselectDropdown.prototype.isSelected = function (option) {
                    return this.model && this.model.indexOf(option.id) > -1;
                };
                MultiselectDropdown.prototype.setSelected = function (event, option) {
                    if (!this.model) {
                        this.model = [];
                    }
                    var index = this.model.indexOf(option.id);
                    if (index > -1) {
                        this.model.splice(index, 1);
                    }
                    else {
                        if (this.settings.selectionLimit === 0 || this.model.length < this.settings.selectionLimit) {
                            this.model.push(option.id);
                        }
                        else {
                            if (this.settings.autoUnselect) {
                                this.model.push(option.id);
                                this.model.shift();
                            }
                            else {
                                this.selectionLimitReached.emit(this.model.length);
                                return;
                            }
                        }
                    }
                    if (this.settings.closeOnSelect) {
                        this.toggleDropdown();
                    }
                    this.onModelChange(this.model);
                };
                MultiselectDropdown.prototype.updateNumSelected = function () {
                    this.numSelected = this.model && this.model.length || 0;
                };
                MultiselectDropdown.prototype.updateTitle = function () {
                    var _this = this;
                    if (this.numSelected === 0) {
                        this.title = this.texts.defaultTitle;
                    }
                    else if (this.settings.dynamicTitleMaxItems >= this.numSelected) {
                        this.title = this.options
                            .filter(function (option) {
                            return _this.model && _this.model.indexOf(option.id) > -1;
                        })
                            .map(function (option) { return option.name; })
                            .join(', ');
                    }
                    else {
                        this.title = this.numSelected
                            + ' '
                            + (this.numSelected === 1 ? this.texts.checked : this.texts.checkedPlural);
                    }
                };
                MultiselectDropdown.prototype.checkAll = function () {
                    this.model = this.options.map(function (option) { return option.id; });
                    this.onModelChange(this.model);
                };
                MultiselectDropdown.prototype.uncheckAll = function () {
                    this.model = [];
                    this.onModelChange(this.model);
                };
                return MultiselectDropdown;
            }());
            __decorate([
                core_1.Input(),
                __metadata("design:type", Array)
            ], MultiselectDropdown.prototype, "options", void 0);
            __decorate([
                core_1.Input(),
                __metadata("design:type", Object)
            ], MultiselectDropdown.prototype, "settings", void 0);
            __decorate([
                core_1.Input(),
                __metadata("design:type", Object)
            ], MultiselectDropdown.prototype, "texts", void 0);
            __decorate([
                core_1.Output(),
                __metadata("design:type", Object)
            ], MultiselectDropdown.prototype, "selectionLimitReached", void 0);
            __decorate([
                core_1.Output(),
                __metadata("design:type", Object)
            ], MultiselectDropdown.prototype, "dropdownClosed", void 0);
            __decorate([
                core_1.HostListener('document: click', ['$event.target']),
                __metadata("design:type", Function),
                __metadata("design:paramtypes", [HTMLElement]),
                __metadata("design:returntype", void 0)
            ], MultiselectDropdown.prototype, "onClick", null);
            MultiselectDropdown = __decorate([
                core_1.Component({
                    selector: 'ss-multiselect-dropdown',
                    providers: [MULTISELECT_VALUE_ACCESSOR],
                    styles: ["\n\t   a { outline: none !important; }\n  "],
                    templateUrl: './multiselect-dropdown.html'
                }),
                __metadata("design:paramtypes", [core_1.ElementRef,
                    core_1.IterableDiffers])
            ], MultiselectDropdown);
            exports_1("MultiselectDropdown", MultiselectDropdown);
            MultiselectDropdownModule = (function () {
                function MultiselectDropdownModule() {
                }
                return MultiselectDropdownModule;
            }());
            MultiselectDropdownModule = __decorate([
                core_1.NgModule({
                    imports: [common_1.CommonModule, forms_1.FormsModule],
                    exports: [MultiselectDropdown],
                    declarations: [MultiselectDropdown, MultiSelectSearchFilter],
                })
            ], MultiselectDropdownModule);
            exports_1("MultiselectDropdownModule", MultiselectDropdownModule);
        }
    };
});
